File: python/algo/202606/algoiv08.02.py
from typing import List


class Solution:
    def pathWithObstacles(self, obstacleGrid: List[List[int]]) -> List[List[int]]:
        m, n = len(obstacleGrid), len(obstacleGrid[0])
        ans = []
        seen = [[False] * n for _ in range(m)]
        
        def dfs(r : int, c: int) -> bool:
            if r >= m or c >=n or obstacleGrid[r][c] == 1 or seen[r][c]:
                return False

            seen[r][c] = True
            if (r == m-1 and c == n-1) or dfs(r+1, c) or dfs(r, c+1):
                ans.append([r,c])
                return True
            
            return False
        
        dfs(0, 0)
        return ans[::-1]

File: cplus/leetcode/202606/algoiv08.02.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<vector<int>> pathWithObstacles(vector<vector<int>>& obstacleGrid) {
        // 基本功
        int m = obstacleGrid.size(), n = obstacleGrid[0].size();
        vector seen(m, vector<bool>(n));

        vector<vector<int>> ans;

        auto dfs = [&](this auto&& dfs, int r, int c) -> bool {
            if (r >= m || c >= n || obstacleGrid[r][c] == 1 || seen[r][c]) return false;

            seen[r][c] = true;            
            if ((r == m-1 && c == n-1) || dfs(r+1, c) || dfs(r, c+1)) {
                ans.push_back({r, c}); return true;
            }

            return false;
        };
        dfs(0, 0);
        reverse(ans.begin(), ans.end());
        return ans;     
    }
};